### ISSUE TITLE
[IMPROVEMENT] Support K8s 1.25 by updating removed deprecated resource versions like PodSecurityPolicy

### ISSUE THREAD

**Issue by innobead** [longhorn/longhorn#4003]
## Is your improvement request related to a feature? Please describe

PodSecurityPolicy has been deprecated and will be removed from K8s 1.25, so we need to find an alternative way to resolve the need for PSP in Longhorn to support 1.25.

Also, some deprecated resource versions are also removed from 1.25. Need to resolve this via https://github.com/longhorn/longhorn/issues/4239 or even consider to backport this to 1.3 & 1.2 via an adaptive way to determine the K8s version of the cluster to use which API resource version if possible (except PSP, because it's totally removed instead of version bump). 

- Cronjob v1beta1 -> v1
- EndpointSlice v1beta1 -> v1
- Event v1beta1 -> v1
- HorizontalPodAutoscaler v2beta1 -> v2
- PodDisruptionBudget v1beta1 -> removed

Note: client-go is backward compatible with K8s any version.

> Compatibility: client-go <-> Kubernetes clusters
> Since Kubernetes is backwards compatible with clients, older client-go versions will work with many different Kubernetes cluster versions.

## Describe the solution you'd like

Deprecate PSP if it's not needed. Otherwise, we need an alternative solution like https://kubernetes.io/docs/tasks/configure-pod-container/migrate-from-psp/.

## Describe alternatives you've considered

N/A

## Additional context

https://www.kubernetes.dev/resources/release/#timeline
https://kubernetes.io/docs/reference/using-api/deprecation-guide/#v1-25
https://github.com/longhorn/longhorn/issues/4239

**Comment by innobead** [comment]
This needs to consider to backport.

**Comment by mattmeye** [comment]
This was released in [Longhorn v1.2.5-rc1 release](https://github.com/longhorn/longhorn/releases/tag/v1.2.5-rc1) but not in [Longhorn v1.2.5 release](https://github.com/longhorn/longhorn/releases/tag/v1.2.5). Is there an ETA for an official release?

**Comment by innobead** [comment]
> This was released in [Longhorn v1.2.5-rc1 release](https://github.com/longhorn/longhorn/releases/tag/v1.2.5-rc1) but not in [Longhorn v1.2.5 release](https://github.com/longhorn/longhorn/releases/tag/v1.2.5). Is there an ETA for an official release?

It was wrongly mentioned in the rc release note. It is planned for 1.4.0 instead, the end of this year.

**Comment by innobead** [comment]
> dont' forget CronJobs [#4490 (comment)](https://github.com/longhorn/longhorn/issues/4490#issuecomment-1240400365)

Thanks :+1: , all deprecated resources will be taken care of together.

**Comment by skandragon** [comment]
I managed to upgrade to kubernetes 1.25 on several nodes before I noticed the issue, so I'm hoping a release comes out that will handle that version as rolling back isn't an option.

**Comment by mattmeye** [comment]
@skandragon I have exactly the same problem too. However, I cannot wait until the end of the year and unfortunately have to set up the cluster again.

**Comment by PurpleBooth** [comment]
> It was wrongly mentioned in the rc release note. It is planned for 1.4.0 instead, the end of this year.

Are you sure about this scheduling, by the end of the year 1.26 (Tuesday 6th December 2022) will be out?

**Comment by innobead** [comment]
Knowing people are eager for this, but for now, the plan is still the same, because Longhorn has its release scope and cadence. 

Please keep following up on the update here, and we will work on this soon or later.

**Comment by skandragon** [comment]
@innobead what is required to do beyond just changing the API calls?  I know that means version 1.3 would be locked to 1.25 or higher; the better option is likely to try to use the proper API calls depending on what version was in use, I suppose?

I know I was able to just change the betav1 magic to v1 and it just worked as far as I could tell.

A bigger issue for me with longhorn is unrelated, and that's that it won't run under Talos.  :/

**Comment by BjoernPetersen** [comment]
Just want to mention that the wording of the issue is misleading.

> PodSecurityPolicy will be deprecated from K8s 1.25 [...]

PodSecurityPolicy has been deprecated since K8s 1.21 and will be **removed** in 1.25.

**Comment by kjaleshire** [comment]
That's really unfortunate, Longhorn will be unusable by the most recent Kubernetes release for the next two months.

I encourage you to reconsider this release plan.

**Comment by dominch** [comment]
I just hit this issue on k3s (latest channel) and needed to revert to 1.24 (where other coredns bug is present) 
This should happen ASAP because 1.25 is on the way and any service based on longhorn will fail :(

**Comment by PhanLe1010** [comment]
### Why we need PSP before?

If the cluster admin enables PSP admission controller, the default behavior was to block any pod that doesn't have a PSP associated with them. That was the reason Longhorn ship the [longhorn-psp](https://github.com/longhorn/longhorn/blob/master/chart/templates/psp.yaml)

So, now that PSP is removed, we can just remove the PSP template. Similar to how Calico did https://github.com/projectcalico/calico/pull/6270

### Kubernetes provide a new mechanism, [Pod Security Admission](https://kubernetes.io/docs/concepts/security/pod-security-admission/). Does Longhorn need to do anything?

* The way that Pod Security Admission works is by having users to annotate the target namespace with annotations to specify the blocking criteria, [example](https://kubernetes.io/docs/tasks/configure-pod-container/enforce-standards-namespace-labels/). By default, it doesn't block pods in a namespace, [link](https://kubernetes.io/docs/tasks/configure-pod-container/enforce-standards-admission-controller/)
* If the users change the default behavior to block privileged pods, it is the responsibility of the users to add the correct annotation to the namespace that is running Longhorn pods to allow Longhorn pods to start (because Longhorn pods require privileged access to manage storage). Longhorn doesn't manage namespace resources so Longhorn doesn't need to take action here.
* In conclusion, Longhorn doesn't need to take action here other than document that `If user enabled Pod Security Admission and users change the default behavior to block privileged pods, the users to add the correct annotation to the namespace that is running Longhorn pods to allow Longhorn pods to start (because Longhorn pods require privileged access to manage storage)`

**Comment by PhanLe1010** [comment]
### Test plan:
Case 1: in Kubernetes 1.24 with PSP admission controller enabled
* Remove `longhorn-psp` and deploy Longhorn, Longhorn pods should fail to come up
* Add `longhorn-psp` back, Longhorn pods should fail to come up

Case 2: in Kubernetes 1.25 with Pod Security Admission enabled
* Remove `longhorn-psp` template
* Deploy Longhorn, Longhorn pods should come up

Case 3: in Kubernetes 1.25 with Pod Security Admission enabled
* Remove `longhorn-psp` template
* Change the default behavior of Pod Security Admission to block privileged pods
* Deploy Longhorn, Longhorn pods should fail to come up

Case 4: in Kubernetes 1.25 with Pod Security Admission enabled
* Remove `longhorn-psp` template
* Change the default behavior of Pod Security Admission to block privileged pods
* Annotate `longhorn-system` namespace with correct permission
* Deploy Longhorn, Longhorn pods should fail to come up

**Comment by yqlbu** [comment]
I need this ASAP as well :)

**Comment by innobead** [comment]
Why this can not be backported to 1.3 or 1.2? 

Basically, it's because CronJob and PSP v1 are only available since 1.21, but batch/v1beta1 and policy/v1beta1 have been removed from 1.25. So, if backport this fix, that means we need to update the K8s min version for existing release branches 1.3 and 1.2 from 1.18 to 1.21; However, this will not conform to our release branch policy w/ the same K8s version range for the following patch releases of already released branches.

@PhanLe1010 Please correct me if anything is missing or incorrect.

**Comment by warrenrees** [comment]
> 

I apologize for the potentially stupid statement/solution below, I don't really know GO all that well, however wouldn't you be able to detect what version of kubernetes longhorn is running on and use the appropriate resource versions?

Something along the lines of 

if kubernetes.version >= 1.25 {
  imPDBs = make(map[string]*policyv1.PodDisruptionBudget)
} else {
  imPDBs = make(map[string]*policyv1beta1.PodDisruptionBudget)
}

**Comment by PhanLe1010** [comment]
@warrenrees I wish it was that simple. However, we have controllers with listers/informers that notify and cache the objects into the local caches. For example, for the cronjob it is https://github.com/longhorn/longhorn-manager/blob/15f850c6937a1eedb5667dfc775c1b7037b45b9c/datastore/datastore.go#L70-L71. I am not sure how to easily/cleanly solve this dynamic multi-typed problem.

**Comment by warrenrees** [comment]
Yeah I see the conundrum... short of adding another entry into the DataStore struct (cjListerV1   batchlisters_v1.CronJobLister) and some logic in the NewDataStore function, it gets complicated fairly quickly.

Sorry just trying to brainstorm how to fix this without having to wait 2 months.  I have 2 of 5 nodes down in my K8s cluster and I don't particularly want to rebuild it :)

Thank you guys for all your hard work, I love longhorn it's a great storage solution!

**Comment by MohammedNoureldin** [comment]
Are there any updates? I read enough to start with Longhorn, and then that disappointing moment came, I cannot use it with 1.25 K8s.

**Comment by PhanLe1010** [comment]
Is it for testing or production use @MohammedNoureldin ?
This one will take sometime until we release Longhorn v1.4.0. But if you are just testing, you can use master-head `kubectl apply -f https://raw.githubusercontent.com/longhorn/longhorn/master/deploy/longhorn.yaml`

**Comment by MohammedNoureldin** [comment]
Thanks for the almost immediate reply, @PhanLe1010!

Is there any nightly built Helm chart to be installed?

**Comment by PhanLe1010** [comment]
Here is it https://github.com/longhorn/longhorn/tree/master/chart 

This chart will track `master-head` which is the latest and unstable version.

**Comment by MohammedNoureldin** [comment]
> (subjected to change any time)

I know that you mean to an earlier date ;)

Maybe not really relevant question, but adding a GitHub repo and installing this chart is not really something that works out of the box, and no link online helped so far. Could you summarize that in 2 lines of bash commands?

**Comment by PhanLe1010** [comment]
The bash command would be like
`git clone https://github.com/longhorn/longhorn.git && cd longhorn && helm install longhorn ./chart --namespace longhorn-system --create-namespace`

Another alternative is `kubectl apply -f https://raw.githubusercontent.com/longhorn/longhorn/master/deploy/longhorn.yaml`

**Comment by MohammedNoureldin** [comment]
Oh, thanks! I thought it is possible to install it using Helm from a GitHub repo on the fly, but it turned out that I need to clone it. I will try it now. Thanks a lot!

**Comment by dkrizic** [comment]
I just want to confirm that the helm chart @ master works excellently on Kubernetes 1.25.3 with the default settings. I am now waiting for the release 1.4.0

**Comment by chriscchien** [comment]
### **Case 1: within a Kubernetes version**
- [ ] **Test basic recurring snapshot/backup jobs are working correctly**

| Kubernetes Version | Pipeline Link | Verify Result | 
|---|---|---|
|v1.25.3+k3s1|Verified on local environment|:heavy_check_mark:|
|v1.24.7+k3s1|https://ci.longhorn.io/job/private/job/longhorn-tests-regression/2239/|In Progress|
|v1.23.1+k3s2|https://ci.longhorn.io/job/public/job/master/job/sles/job/amd64/job/longhorn-tests-sles-amd64/292/testReport/tests/test_recurring_job/|:heavy_check_mark:|
|v1.22.15+k3s1|https://ci.longhorn.io/job/private/job/longhorn-tests-regression/2238/|In Progress|
|v1.21.10+k3s1|https://ci.longhorn.io/job/private/job/longhorn-tests-regression/2237/|In Progress|

<br/>

- [x] **Test generating support bundle generation. Make sure that the above jobs are in /yamls/kubernetes/cronjobs folder**

| Kubernetes Version | Verify Result | 
|---|---|
|v1.25.3+k3s1|:heavy_check_mark:|
|v1.24.7+k3s1|:heavy_check_mark:|
|v1.23.13+k3s1|:heavy_check_mark:|
|v1.22.15+k3s1|:heavy_check_mark:|
|v1.21.10+k3s1|:heavy_check_mark:|
 
<br/>

- [ ] **Test upgrade Longhorn from 1.3.2 or v1.2.5 to master**

Test basic functionalities (workload, upgrade engine image, replica rebuilding)

   | Kubernetes Version | Upgrade from v1.3.2 | Upgrade from v1.2.5 | 
   |---|---|---|
   |v1.25.3+k3s1|NA|NA|
   |v1.24.7+k3s1|:heavy_check_mark:|:heavy_check_mark:|
   |v1.23.13+k3s1|:heavy_check_mark:|:heavy_check_mark:|
   |v1.22.15+k3s1|:heavy_check_mark:|:heavy_check_mark:|
   |v1.21.10+k3s1|:heavy_check_mark:|:heavy_check_mark:|

Verify CSI snapshot upgrade

   | Kubernetes Version | Verify Result | 
   |---|---|
   |v1.25.3+k3s1|NA|
   |v1.24.7+k3s1||
   |v1.23.13+k3s1||
   |v1.22.15+k3s1||
   |v1.21.10+k3s1||

<br/>

- [x] **Test zone and topology,**

| Kubernetes Version | Verify Result | 
|---|---|
|v1.25.3+k3s1|:heavy_check_mark:|
|v1.24.7+k3s1|:heavy_check_mark:|
|v1.23.13+k3s1|:heavy_check_mark:|
|v1.22.15+k3s1|:heavy_check_mark:|
|v1.21.10+k3s1|:heavy_check_mark:|

<br/>

- [x] **Test uninstallation**

 | Kubernetes Version | Verify Result | 
|---|---|
|v1.25.3+k3s1|:heavy_check_mark:|
|v1.24.7+k3s1|:heavy_check_mark:|
|v1.23.13+k3s1|:heavy_check_mark:|
|v1.22.15+k3s1|:heavy_check_mark:|
|v1.21.10+k3s1|:heavy_check_mark:|
 
<br/>

### **Case 2: Test a Kubernetes version upgrade**

- [x] Update from `v1.24.7+k3s1` to `v1.25.3+k3s1` and workload still working properly

1. Install Longhorn master in Kubernetes v1.24 cluster
2. Create some workload, Longhorn volumes, cronjobs, etc...
3. Upgrade Kubernetes to v.125
4. Verify that Longhorn is running ok. Basic functionality is working ok

**Comment by khushboo-rancher** [comment]
@chriscchien Could you just test one more scenario, deploying Longhorn on a hardened cluster?

**Comment by abderraxim** [comment]
Head Chart installs perfectly and everything looks fine

**Comment by nabeelshaikh7** [comment]
and some how installing longhorn crashed treafik in my k3s cluster

**Comment by chriscchien** [comment]
Hi @khushboo-rancher ,

I tried deploy Longhorn master on hardened cluster but the deploy not scuccessed.
Deployment `longhorn-driver-deployer` raised error and most Longhorn component pod not deployed 

```
[FailedCreate] pods "longhorn-driver-deployer-f9546bd9b-" is forbidden: PodSecurityPolicy: unable to admit pod: [spec.initContainers[0].securityContext.runAsUser: Invalid value: 0: running with the root UID is forbidden spec.containers[0].securityContext.runAsUser: Invalid value: 0: running with the root UID is forbidden]
```

**Comment by yangchiu** [comment]
> Hi @khushboo-rancher ,
> 
> I tried deploy Longhorn master on hardened cluster but the deploy not scuccessed. Deployment `longhorn-driver-deployer` raised error and most Longhorn component pod not deployed
> 
> ```
> [FailedCreate] pods "longhorn-driver-deployer-f9546bd9b-" is forbidden: PodSecurityPolicy: unable to admit pod: [spec.initContainers[0].securityContext.runAsUser: Invalid value: 0: running with the root UID is forbidden spec.containers[0].securityContext.runAsUser: Invalid value: 0: running with the root UID is forbidden]
> ```

@chriscchien Would you mind to share the test steps here?

**Comment by PhanLe1010** [comment]
Probably need to enable the PSP on this hardened cluster. Switch this field to `true` https://github.com/longhorn/longhorn/blob/23d264196f6bfdc4928b333e49b30b6c29ad417e/chart/values.yaml#L253-L255

**Comment by chriscchien** [comment]
Using helm to install Longhorn master head by command `helm install longhorn ./chart -n longhorn-system` which `enablePSP` set to `true` can deploy/uninstall Longhorn and workload worked well on the hardened cluster

But there is no PSP field exist in master-head's [longhorn.yaml](https://github.com/longhorn/longhorn/blob/master/deploy/longhorn.yaml), does this mean currently we can not use kubectl to deploy Longhorn on hardened cluster? thank you

cc @khushboo-rancher,  @yangchiu

**Comment by PhanLe1010** [comment]
[longhorn.yaml](https://github.com/longhorn/longhorn/blob/master/deploy/longhorn.yaml) is generated by running [helm template](https://github.com/longhorn/longhorn/blob/23d264196f6bfdc4928b333e49b30b6c29ad417e/scripts/generate-longhorn-yaml.sh#L26) . The default Helm value `enablePSP` is false so it Helm template does not generate PSP yaml into longhorn.yaml. This behavior is expected because we are moving away from PSP. However, it is a bit incovinent for the user using old hardened cluster with PSP enable, therefore we have 2 options:
1. Document and pointing the user to PSP yaml
2. Put the PSP yaml under comment section into the longhorn.yaml. This method would require us to modify the [generate-longhorn-yaml.sh](https://github.com/longhorn/longhorn/blob/master/scripts/generate-longhorn-yaml.sh#L26)

**Comment by PhanLe1010** [comment]
> For use kubectl to deploy Longhorn, I tried add below section into longhon.yaml to deploy but still have problem in longhorn-driver-deployer

You also need to add PSP into `longhorn-service-account`. AKA, you need to put [this whole section](https://github.com/longhorn/longhorn-manager/blob/master/deploy/install/01-prerequisite/05-podsecuritypolicy.yaml) into `longhorn.yaml`

**Comment by chriscchien** [comment]
Thank you @PhanLe1010 ,

Deploy Longhorn on hardened cluster by `kubectl` and `helm` all worked well( including steps install, create/scale workload, and uninstall).

**Comment by innobead** [comment]
> [longhorn.yaml](https://github.com/longhorn/longhorn/blob/master/deploy/longhorn.yaml) is generated by running [helm template](https://github.com/longhorn/longhorn/blob/23d264196f6bfdc4928b333e49b30b6c29ad417e/scripts/generate-longhorn-yaml.sh#L26) . The default Helm value `enablePSP` is `false` so Helm template does not generate PSP yaml into longhorn.yaml. This behavior is expected because we are moving away from PSP. However, it is a bit inconvenient for the user using old hardened clusters with PSP enabled, therefore we have 2 options:
> 
> 1. Document and pointing the user to PSP yaml
> 2. Put the PSP yaml under the comment section into the `longhorn.yaml`. This method would require us to modify the [generate-longhorn-yaml.sh](https://github.com/longhorn/longhorn/blob/master/scripts/generate-longhorn-yaml.sh#L26)
> 
> cc @innobead Which option would you suggest from PM perspective?

I think we need to do both for users who use kubectl to apply the manifest directly. If in the future, it becomes complicated, we can use kustomize instead, but right now it's enough to have a separate manifest for < 1.25. 

@PhanLe1010 Could you help with this? Thanks.

**Comment by chriscchien** [comment]
In addition, tested upgrade Longhorn from old release to v1.40-rc1 then upgrade Kubernetes version to v1.25 
Below scenario were all **passed**

In v1.24.7+k3s1

- [ ] Upgrade Longhorn from v1.2.6 to v1.4.0-rc1 then upgrade kubernetes to  v1.25.4+k3s1
- [ ] Upgrade Longhorn from v1.3.2 to v1.4.0-rc1 then upgrade kubernetes to v1.25.4+k3s1

In v1.23.11+k3s1

- [ ] Upgrade Longhorn from v1.2.6 to v1.4.0-rc1 then upgrade kubernetes to v1.25.4+k3s1
- [ ] Upgrade Longhorn from v1.3.2 to v1.4.0-rc1 then upgrade kubernetes to v1.25.4+k3s1
